src/utils/graceful_shutdown.rs

```diff
@@ -0,0 +1,208 @@
+//! A class to handle shutdown signals and to shut down system
+//! Actix system has to be running for it to work.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    mem,
+    time::Duration,
+};
+
+use actix::{
+    self,
+    prelude::{Actor, Context},
+    Addr, AsyncContext, Handler, Message, Recipient, ResponseActFuture, System,
+    WrapFuture,
+};
+
+use tokio::prelude::{
+    future::{self, join_all, Future},
+    stream::*,
+    FutureExt,
+};
+
+use crate::log::prelude::*;
+
+pub type ShutdownMessageResult =
+    Result<Box<(dyn Future<Item = (), Error = ()> + std::marker::Send)>, ()>;
```

Comment:
@tyranron,

`Send` is required here because we might (and probably will, since we are launching a new arbiter for shutdown) send this future to another thread

Reply:
@tyranron,

`Send` is required here because we might (and probably will, since we are creating new arbiter), send this future to a different thread

Reply:
@nWacky are you sure about that? AFAIK, actix allows to pass stuff between `Actor`s without explicit `Send` requirement. Could you specify a link to API, which requires this?